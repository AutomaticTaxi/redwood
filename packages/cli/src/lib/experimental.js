import path from 'path'

import { getConfig, getPaths, resolveFile } from '@redwoodjs/project-config'

/**
 * Checks to see if the user has the experimental server file present at api/src/server.ts|js and enabled in redwood.toml
 *
 * @returns {boolean} true if the experimental server file is enabled and present
 */
export function isUsingExperimentalServerFile() {
  return (
    getConfig().experimental?.serverFile?.enabled &&
    resolveFile(path.join(getPaths().api.src, 'server'))
  )
}
