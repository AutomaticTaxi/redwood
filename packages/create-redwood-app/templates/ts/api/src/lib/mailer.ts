import { Mailer } from '@redwoodjs/mailer-core'
import { InMemoryMailHandler } from '@redwoodjs/mailer-handler-in-memory'
import { NodemailerMailHandler } from '@redwoodjs/mailer-handler-nodemailer'
import { StudioMailHandler } from '@redwoodjs/mailer-handler-studio'
import { ReactEmailRenderer } from '@redwoodjs/mailer-renderer-react-email'

import { logger } from './logger'

export const mailer = new Mailer({
  // Handlers to send our emails out to the world
  handling: {
    handlers: {
      inMemory: new InMemoryMailHandler(),
      studio: new StudioMailHandler(),
      // TODO: Update this handler config or switch it out for a different handler completely
      nodemailer: new NodemailerMailHandler({
        transport: {
          host: 'localhost',
          port: 4319,
          secure: false,
        },
      }),
    },
    // This is our default handler when in production only
    default: 'nodemailer',
  },

  // Renderers to convert our templates into HTML/text
  rendering: {
    renderers: {
      reactEmail: new ReactEmailRenderer(),
    },
    default: 'reactEmail',
  },

  // Specific handlers for testing and development time
  test: {
    handler: 'inMemory',
    // when: process.env.NODE_ENV === 'test',
  },
  development: {
    handler: 'studio',
    // when: process.env.NODE_ENV !== 'production',
  },

  // Default send options
  defaults: {
    // replyTo: 'noreply@example.com',
  },

  logger,
})
