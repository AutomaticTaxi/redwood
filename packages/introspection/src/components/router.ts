import { getPaths } from '@redwoodjs/project-config'

import type {
  RedwoodIntrospectionError,
  RedwoodIntrospectionWarning,
} from './introspection'
import { RedwoodIntrospectionComponent } from './introspection'
import { RedwoodRoute } from './route'

export class RedwoodRouter extends RedwoodIntrospectionComponent {
  readonly type = 'router'

  readonly routes: RedwoodRoute[]

  private constructor(filepath: string) {
    super(filepath)

    this.routes = RedwoodRoute.parseRoutes(this)
  }

  getErrors(): RedwoodIntrospectionError[] {
    const errors: RedwoodIntrospectionError[] = []

    // Not found route checks
    const notFoundRoutes = this.routes.filter((route) => {
      return route.isNotFound
    })
    if (notFoundRoutes.length === 0) {
      errors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'No "notfound" route detected',
      })
    } else if (notFoundRoutes.length > 1) {
      errors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'No more than one "notfound" route should be present',
      })
    }

    // Duplicate routes checking
    const nameOccurences: Record<string, number> = {}
    this.routes.forEach((route) => {
      if (route.name in nameOccurences && nameOccurences[route.name] < 2) {
        errors.push({
          component: {
            type: this.type,
            filepath: this.filepath,
            name: this.name,
          },
          message: `Multiple routes named "${route.name}" are present`,
        })
        nameOccurences[route.name] += 1
      } else {
        nameOccurences[route.name] = 1
      }
    })

    return errors
  }

  getWarnings(): RedwoodIntrospectionWarning[] {
    const warnings: RedwoodIntrospectionWarning[] = []

    if (this.routes.length === 0) {
      warnings.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: 'No routes found',
      })
    }

    return warnings
  }

  static parseRouter(filepath = getPaths().web.routes) {
    return new RedwoodRouter(filepath)
  }
}
