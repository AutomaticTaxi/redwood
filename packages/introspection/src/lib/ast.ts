import fs from 'node:fs'
import path from 'node:path'

import { parse as babelParse } from '@babel/parser'
import type { ParserPlugin } from '@babel/parser'
import traverse from '@babel/traverse'
import type {
  JSXElement,
  ExportNamedDeclaration,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/project-config'

export function getASTFromFile(filepath: string) {
  const code = fs.readFileSync(filepath, { encoding: 'utf8', flag: 'r' })

  // use jsx plugin for web files, because in JS, the .jsx extension is not used
  const isJsxFile =
    path.extname(filepath).match(/[jt]sx$/) ||
    path.parse(filepath).dir.startsWith(getPaths().web.base)

  const plugins = [
    'typescript',
    'nullishCoalescingOperator',
    'objectRestSpread',
    isJsxFile && 'jsx',
  ].filter(Boolean) as ParserPlugin[]

  return babelParse(code, {
    sourceType: 'module',
    plugins,
  })
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(element: JSXElement) {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
