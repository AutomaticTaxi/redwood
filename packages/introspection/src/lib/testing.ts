import { ensurePosixPath } from '@redwoodjs/project-config'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
