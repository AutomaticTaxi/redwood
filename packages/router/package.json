{
  "name": "@redwoodjs/router",
  "version": "7.0.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/redwoodjs/redwood.git",
    "directory": "packages/router"
  },
  "license": "MIT",
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./skip-nav.css": "./skip-nav.css",
    "./dist/ActivePageContext": {
      "types": "./dist/ActivePageContext.d.ts",
      "import": "./dist/ActivePageContext.js",
      "default": "./dist/cjs/ActivePageContext.js"
    },
    "./dist/AuthenticatedRoute": {
      "types": "./dist/AuthenticatedRoute.d.ts",
      "import": "./dist/AuthenticatedRoute.js",
      "default": "./dist/cjs/AuthenticatedRoute.js"
    },
    "./dist/PageLoadingContext": {
      "types": "./dist/PageLoadingContext.d.ts",
      "import": "./dist/PageLoadingContext.js",
      "default": "./dist/cjs/PageLoadingContext.js"
    },
    "./dist/Route": {
      "types": "./dist/Route.d.ts",
      "import": "./dist/Route.js",
      "default": "./dist/cjs/Route.js"
    },
    "./dist/Set": {
      "types": "./dist/Set.d.ts",
      "import": "./dist/Set.js",
      "default": "./dist/cjs/Set.js"
    },
    "./dist/a11yUtils": {
      "types": "./dist/a11yUtils.d.ts",
      "import": "./dist/a11yUtils.js",
      "default": "./dist/cjs/a11yUtils.js"
    },
    "./dist/active-route-loader": {
      "types": "./dist/active-route-loader.d.ts",
      "import": "./dist/active-route-loader.js",
      "default": "./dist/cjs/active-route-loader.js"
    },
    "./dist/analyzeRoutes": {
      "types": "./dist/analyzeRoutes.d.ts",
      "import": "./dist/analyzeRoutes.js",
      "default": "./dist/cjs/analyzeRoutes.js"
    },
    "./dist/createNamedContext": {
      "types": "./dist/createNamedContext.d.ts",
      "import": "./dist/createNamedContext.js",
      "default": "./dist/cjs/createNamedContext.js"
    },
    "./dist/dummyComponent": {
      "types": "./dist/dummyComponent.d.ts",
      "import": "./dist/dummyComponent.js",
      "default": "./dist/cjs/dummyComponent.js"
    },
    "./dist/history": {
      "types": "./dist/history.d.ts",
      "import": "./dist/history.js",
      "default": "./dist/cjs/history.js"
    },
    "./dist/index": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./dist/link": {
      "types": "./dist/link.d.ts",
      "import": "./dist/link.js",
      "default": "./dist/cjs/link.js"
    },
    "./dist/location": {
      "types": "./dist/location.d.ts",
      "import": "./dist/location.js",
      "default": "./dist/cjs/location.js"
    },
    "./dist/namedRoutes": {
      "types": "./dist/namedRoutes.d.ts",
      "import": "./dist/namedRoutes.js",
      "default": "./dist/cjs/namedRoutes.js"
    },
    "./dist/navLink": {
      "types": "./dist/navLink.d.ts",
      "import": "./dist/navLink.js",
      "default": "./dist/cjs/navLink.js"
    },
    "./dist/page": {
      "types": "./dist/page.d.ts",
      "import": "./dist/page.js",
      "default": "./dist/cjs/page.js"
    },
    "./dist/params": {
      "types": "./dist/params.d.ts",
      "import": "./dist/params.js",
      "default": "./dist/cjs/params.js"
    },
    "./dist/redirect": {
      "types": "./dist/redirect.d.ts",
      "import": "./dist/redirect.js",
      "default": "./dist/cjs/redirect.js"
    },
    "./dist/route-announcement": {
      "types": "./dist/route-announcement.d.ts",
      "import": "./dist/route-announcement.js",
      "default": "./dist/cjs/route-announcement.js"
    },
    "./dist/route-focus": {
      "types": "./dist/route-focus.d.ts",
      "import": "./dist/route-focus.js",
      "default": "./dist/cjs/route-focus.js"
    },
    "./dist/route-validators": {
      "types": "./dist/route-validators.d.ts",
      "import": "./dist/route-validators.js",
      "default": "./dist/cjs/route-validators.js"
    },
    "./dist/routeParamsTypes": {
      "types": "./dist/routeParamsTypes.d.ts",
      "import": "./dist/routeParamsTypes.js",
      "default": "./dist/cjs/routeParamsTypes.js"
    },
    "./dist/router-context": {
      "types": "./dist/router-context.d.ts",
      "import": "./dist/router-context.js",
      "default": "./dist/cjs/router-context.js"
    },
    "./dist/router": {
      "types": "./dist/router.d.ts",
      "import": "./dist/router.js",
      "default": "./dist/cjs/router.js"
    },
    "./dist/rsc-link": {
      "types": "./dist/rsc-link.d.ts",
      "import": "./dist/rsc-link.js",
      "default": "./dist/cjs/rsc-link.js"
    },
    "./dist/server-route-loader": {
      "types": "./dist/server-route-loader.d.ts",
      "import": "./dist/server-route-loader.js",
      "default": "./dist/cjs/server-route-loader.js"
    },
    "./dist/server-router": {
      "types": "./dist/server-router.d.ts",
      "import": "./dist/server-router.js",
      "default": "./dist/cjs/server-router.js"
    },
    "./dist/skipNav": {
      "types": "./dist/skipNav.d.ts",
      "import": "./dist/skipNav.js",
      "default": "./dist/cjs/skipNav.js"
    },
    "./dist/splash-page": {
      "types": "./dist/splash-page.d.ts",
      "import": "./dist/splash-page.js",
      "default": "./dist/cjs/splash-page.js"
    },
    "./dist/useIsMounted": {
      "types": "./dist/useIsMounted.d.ts",
      "import": "./dist/useIsMounted.js",
      "default": "./dist/cjs/useIsMounted.js"
    },
    "./dist/useMatch": {
      "types": "./dist/useMatch.d.ts",
      "import": "./dist/useMatch.js",
      "default": "./dist/cjs/useMatch.js"
    },
    "./dist/useRouteName": {
      "types": "./dist/useRouteName.d.ts",
      "import": "./dist/useRouteName.js",
      "default": "./dist/cjs/useRouteName.js"
    },
    "./dist/useRoutePaths": {
      "types": "./dist/useRoutePaths.d.ts",
      "import": "./dist/useRoutePaths.js",
      "default": "./dist/cjs/useRoutePaths.js"
    },
    "./dist/util": {
      "types": "./dist/util.d.ts",
      "import": "./dist/util.js",
      "default": "./dist/cjs/util.js"
    }
  },
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "skip-nav.css"
  ],
  "scripts": {
    "build": "tsx ./build.mts && yarn build:types",
    "build:pack": "yarn pack -o redwoodjs-router.tgz",
    "build:types": "tsc --build --verbose tsconfig.build.json",
    "build:watch": "nodemon --watch src --ext \"js,jsx,ts,tsx\" --ignore dist --exec \"yarn build\"",
    "prepublishOnly": "NODE_ENV=production yarn build",
    "test": "vitest run",
    "test:types": "tstyche",
    "test:watch": "vitest watch"
  },
  "dependencies": {
    "@redwoodjs/auth": "workspace:*"
  },
  "devDependencies": {
    "@redwoodjs/framework-tools": "workspace:*",
    "@testing-library/jest-dom": "6.4.5",
    "@types/react": "^18.2.55",
    "@types/react-dom": "^18.2.19",
    "react": "19.0.0-beta-04b058868c-20240508",
    "react-dom": "19.0.0-beta-04b058868c-20240508",
    "tstyche": "1.1.0",
    "tsx": "4.10.3",
    "typescript": "5.4.5",
    "vitest": "1.6.0"
  },
  "peerDependencies": {
    "react": "19.0.0-beta-04b058868c-20240508",
    "react-dom": "19.0.0-beta-04b058868c-20240508"
  },
  "gitHead": "3905ed045508b861b495f8d5630d76c7a157d8f1"
}
