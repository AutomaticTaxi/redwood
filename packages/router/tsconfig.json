{
  "extends": "../../tsconfig.compilerOption.json",
  "compilerOptions": {
    "isolatedModules": true,
    "moduleResolution": "NodeNext",
    "module": "NodeNext",
    "rootDir": "src",
    "outDir": "dist",
  },
  "include": ["src", "./ambient.d.ts"],
  "references": [
    { "path": "../auth/tsconfig.build.json" },
    { "path": "../framework-tools" },
  ]
}
