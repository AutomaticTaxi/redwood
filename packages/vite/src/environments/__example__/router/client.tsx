"use client"

import React from 'react'

type RouterContextType = {
  path?: string;
}

export const RouterContext = React.createContext<RouterContextType>(undefined!);

export function useRouter() {
  return React.use(RouterContext)
}