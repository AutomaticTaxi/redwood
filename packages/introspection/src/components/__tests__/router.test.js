import path from 'node:path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodRouter } from '../router'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)
const FIXTURE_PROJECTS = [
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
]

describe.each(FIXTURE_PROJECTS)(
  "From within the '%s' fixture",
  (PROJECT_NAME) => {
    // Setup RWJS_CWD to point to the fixture project
    const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)
    const RWJS_CWD = process.env.RWJS_CWD
    beforeAll(() => {
      process.env.RWJS_CWD = PROJECT_PATH
    })
    afterAll(() => {
      process.env.RWJS_CWD = RWJS_CWD
    })

    it('returns the correct router', () => {
      const router = RedwoodRouter.parseRouter()
      router.filepath = stripAndFormatPathForTesting(
        router.filepath,
        PROJECT_PATH
      )
      for (let i = 0; i < router.routes.length; i++) {
        const route = router.routes[i]
        route.filepath = stripAndFormatPathForTesting(
          route.filepath,
          PROJECT_PATH
        )
      }

      expect(router).toMatchSnapshot('router content')
      expect(router.getErrors()).toMatchSnapshot(`${router.filepath} errors`)
      expect(router.getWarnings()).toMatchSnapshot(
        `${router.filepath} warnings`
      )
    })
  }
)
