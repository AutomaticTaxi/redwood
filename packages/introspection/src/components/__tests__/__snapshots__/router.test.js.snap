// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the 'empty-project' fixture returns the correct router: /web/src/Routes.tsx errors 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct router: /web/src/Routes.tsx warnings 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct router: router content 1`] = `
RedwoodRouter {
  "filepath": "/web/src/Routes.tsx",
  "name": "Routes",
  "routes": [
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": true,
      "isPrivate": false,
      "name": "notfound",
      "pageIdentifier": "NotFoundPage",
      "path": undefined,
      "prerender": false,
      "type": "route",
    },
  ],
  "type": "router",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct router: /web/src/Routes.js errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct router: /web/src/Routes.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct router: router content 1`] = `
RedwoodRouter {
  "filepath": "/web/src/Routes.js",
  "name": "Routes",
  "routes": [
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "home",
      "pageIdentifier": "HomePage",
      "path": "/",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "typescriptPage",
      "pageIdentifier": "TypeScriptPage",
      "path": "/typescript",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "someOtherPage",
      "pageIdentifier": "EditUserPage",
      "path": "/somewhereElse",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "fooPage",
      "pageIdentifier": "FooPage",
      "path": "/foo",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "barPage",
      "pageIdentifier": "BarPage",
      "path": "/bar",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": true,
      "name": "privatePage",
      "pageIdentifier": "PrivatePage",
      "path": "/private-page",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": true,
      "isPrivate": false,
      "name": "notfound",
      "pageIdentifier": "NotFoundPage",
      "path": undefined,
      "prerender": false,
      "type": "route",
    },
  ],
  "type": "router",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct router: /web/src/Routes.js errors 1`] = `
[
  {
    "component": {
      "filepath": "/web/src/Routes.js",
      "name": "Routes",
      "type": "router",
    },
    "message": "No "notfound" route detected",
  },
  {
    "component": {
      "filepath": "/web/src/Routes.js",
      "name": "Routes",
      "type": "router",
    },
    "message": "Multiple routes named "home" are present",
  },
]
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct router: /web/src/Routes.js warnings 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct router: router content 1`] = `
RedwoodRouter {
  "filepath": "/web/src/Routes.js",
  "name": "Routes",
  "routes": [
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "home",
      "pageIdentifier": "HomePage",
      "path": "/",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "home",
      "pageIdentifier": "HomePage",
      "path": "/",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "something-else",
      "pageIdentifier": "DoesNotExist",
      "path": "/{foo}/{foo}",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.js",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "Routes",
      "pageIdentifier": "NotFoundPage",
      "path": undefined,
      "prerender": false,
      "type": "route",
    },
  ],
  "type": "router",
}
`;

exports[`From within the 'test-project' fixture returns the correct router: /web/src/Routes.tsx errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct router: /web/src/Routes.tsx warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct router: router content 1`] = `
RedwoodRouter {
  "filepath": "/web/src/Routes.tsx",
  "name": "Routes",
  "routes": [
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "double",
      "pageIdentifier": "DoublePage",
      "path": "/double",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "login",
      "pageIdentifier": "LoginPage",
      "path": "/login",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "signup",
      "pageIdentifier": "SignupPage",
      "path": "/signup",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "forgotPassword",
      "pageIdentifier": "ForgotPasswordPage",
      "path": "/forgot-password",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "resetPassword",
      "pageIdentifier": "ResetPasswordPage",
      "path": "/reset-password",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "newContact",
      "pageIdentifier": "ContactNewContactPage",
      "path": "/contacts/new",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "editContact",
      "pageIdentifier": "ContactEditContactPage",
      "path": "/contacts/{id:Int}/edit",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "contact",
      "pageIdentifier": "ContactContactPage",
      "path": "/contacts/{id:Int}",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "contacts",
      "pageIdentifier": "ContactContactsPage",
      "path": "/contacts",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "newPost",
      "pageIdentifier": "PostNewPostPage",
      "path": "/posts/new",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "editPost",
      "pageIdentifier": "PostEditPostPage",
      "path": "/posts/{id:Int}/edit",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "post",
      "pageIdentifier": "PostPostPage",
      "path": "/posts/{id:Int}",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "posts",
      "pageIdentifier": "PostPostsPage",
      "path": "/posts",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "waterfall",
      "pageIdentifier": "WaterfallPage",
      "path": "/waterfall/{id:Int}",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": true,
      "name": "profile",
      "pageIdentifier": "ProfilePage",
      "path": "/profile",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": true,
      "isNotFound": false,
      "isPrivate": false,
      "name": "blogPost",
      "pageIdentifier": "BlogPostPage",
      "path": "/blog-post/{id:Int}",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "contactUs",
      "pageIdentifier": "ContactUsPage",
      "path": "/contact",
      "prerender": false,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "about",
      "pageIdentifier": "AboutPage",
      "path": "/about",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": false,
      "isPrivate": false,
      "name": "home",
      "pageIdentifier": "HomePage",
      "path": "/",
      "prerender": true,
      "type": "route",
    },
    RedwoodRoute {
      "filepath": "/web/src/Routes.tsx",
      "hasParameters": false,
      "isNotFound": true,
      "isPrivate": false,
      "name": "notfound",
      "pageIdentifier": "NotFoundPage",
      "path": undefined,
      "prerender": true,
      "type": "route",
    },
  ],
  "type": "router",
}
`;
